src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs

```diff
@@ -435,14 +436,98 @@ private X509Certificate2 LoadCertificate(CertificateConfig certInfo, string endp
+                    if (TryReadPemRSAKey(certInfo, environment, out var rsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate =  publicCertificate.CopyWithPrivateKey(rsaKey);
+                    }
+
+                    if (TryReadPemDSAKey(certInfo, environment, out var dsaKey))
+                    {
+                        var publicCertificate = new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path));
+                        certificate = publicCertificate.CopyWithPrivateKey(dsaKey);
+                    }
+
+                    if (certificate != null)
+                    {
+                        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                        {
+                            return PersistKey(certificate);
+                        }
+
+                        return certificate;
+                    }
+
+                    throw new InvalidOperationException(CoreStrings.InvalidPemKey);
+                }
+
+                return new X509Certificate2(Path.Combine(environment.ContentRootPath, certInfo.Path), certInfo.Password);
             }
             else if (certInfo.IsStoreCert)
             {
                 return LoadFromStoreCert(certInfo);
             }
             return null;
+
+            static X509Certificate2 PersistKey(X509Certificate2 fullCertificate)
+            {
+                // We need to force the key to be persisted.
+                // See https://github.com/dotnet/runtime/issues/23749
+                var certificateBytes = fullCertificate.Export(X509ContentType.Pkcs12, "");
+                return new X509Certificate2(certificateBytes, "", X509KeyStorageFlags.DefaultKeySet);
+            }
+        }
+
+        private static bool TryReadPemRSAKey(CertificateConfig certInfo, IHostEnvironment environment, out RSA key)
+        {
+            try
+            {
+                var keyPath = Path.Combine(environment.ContentRootPath, certInfo.KeyPath);
+                var rsaKey = RSA.Create();
+                if (certInfo.Password == null)
+                {
+                    rsaKey.ImportFromPem(File.ReadAllText(keyPath));
+                }
+                else
+                {
+                    rsaKey.ImportFromEncryptedPem(File.ReadAllText(keyPath), certInfo.Password);
+                }
+                key = rsaKey;
+                return true;
+            }
+            catch
+            {
+                key = null;
+                return false;
+            }
+        }
+
+        private static bool TryReadPemDSAKey(CertificateConfig certInfo, IHostEnvironment environment, out DSA dsaKey)
+        {
+            try
+            {
+                var keyPath = Path.Combine(environment.ContentRootPath, certInfo.KeyPath);
+                var dsa = DSA.Create();
```

Comment:
If you continue with the manually-loading-keys approach, you should dispose this in your catch block,